[ignore-files] Dynamically create ignore entries when calling `add_globs()` and `add_file()`

peabnuts123 opened the pull request:
# Summary 
Calls to `add_globs()` and `add_file()` dynamically create a new ignore entry if there isn't one at the location of `applies_in` param. This allows users to e.g. add globs to a path that previously has no ignore files in.

Fixes #907

peabnuts123 commented:
I have a question about this change: Are there concerns for more niche use-cases? For example, adding a glob that specifically applies_in a path that isn't the origin AND doesn't have an ignore file? (e.g. ignoring `*.d.ts` in `src/` or something). 

I am worried this change is just passing the buck such that now, specifically, paths referencing origin will work, but the issue still exists for any other path. Do functions like `add_globs()` et al need some kind of fallback if they can't find a pre-existing ignore at `applies_in` where they create a new one on demand or something?

passcod commented:
Yeah, you're right, it is an incomplete fix. But:
- it's still improving the status quo
- this whole crate is due for a refactor

So I'm happy to merge as-is.

If you run into the pattern you've described later in your own usage, also happy to get a further fix then, but otherwise not a worry at this stage.

peabnuts123 commented:
I agree any improvement is better than none (and it will solve MY issues…). Just while we're talking about it though, thoughts on something like [this](https://github.com/watchexec/watchexec/compare/main...peabnuts123:experimental-changes-adhoc-builder?expand=1) instead? 

Haven't 100% thought it out yet but it seems to work on the ~1 test I have written. If it's going to be ~roughly as much effort but solve the problem properly, it might be worth just doing instead.

passcod commented:
Oh, absolutely! Thanks for taking the time.

Those `if none, then insert` blocks are a bit... I think there's no Entry API in this trie crate that could make this code more pleasant, though.

peabnuts123 commented:
Ah yeah I meant to add a disclaimer I am far from a rust expert. I tried to turn them into some kind of 

```
let builder = (
  ...
)
```

blocks but the lifetime of the newly created builder got away from me. Open to ideas if you have any though. 

I've updated the code in [that branch](https://github.com/watchexec/watchexec/compare/main...peabnuts123:experimental-changes-adhoc-builder?expand=1). Let me know what you think. If you think it's worth merging instead, I can update this PR. Please also let me know if there's some functional or style changes you think it needs. I've only worked on my own personal rust projects thus far, so would love any feedback.

Summary is basically: 
 - In `add_globs()` and `add_file()` (which is broken in the same way - I didn't even mention that in [#907](https://github.com/watchexec/watchexec/issues/907)) insert a new ignore builder on-the-fly if encountering a new path
 - Move tests over into `filterer.rs`, which has a lot more going on, but allowed me to work with some sample ignore files and write tests for the pre-existing positive test-cases also (i.e. ensure this doesn't cause a regression)

passcod commented:
It looks great!

You'll want to run `cargo fmt` to fix the little inconsistencies, but yeah please update this PR with that code.

Thanks again for this work.

peabnuts123 commented:
Hello ago, sorry, I got a bit busy and dropped off the face of the earth for a few days there. I have run `cargo fmt` and updated this PR with those changes!

peabnuts123 commented:
Tests failing on Windows 😞  I will have to pull on my other machine and debug

peabnuts123 commented:
I am not dead!!! Apologies, I have just been giving my main project some priority to get it across its first milestone. I checked this out on my Windows machine and the test failures appear to be legitimate. Unsure why they seem to pass on my Mac, I assume it has to do with Windows vs. Linux paths. Anyway, I've changed some of the logic (I was passing some strings where I was supposed to be passing PathBufs and Rust was being cute about it) and everything is passing on my Windows machine now! Will check on my Mac now, but hopefully everything should be good now.

passcod commented:
Wonderful! And no worries, fully understand prioritising :)

I may be a few days as I'm away from home at the moment, but at a glance at latest changes it's looking good.
